src/compose/app.ts

```diff
@@ -302,23 +358,51 @@ export class App {
 		return outputs;
 	}
 
+	private getDependentServices<T extends Volume | Network>(
+		component: T,
+		dependencyFn: (component: T, service: Service) => boolean,
+	) {
+		return this.services.filter((s) => dependencyFn(component, s));
```

Comment:
This won't work if the service is in both current and target and needs to be killed to kill a network/volume it's depending on. In that case, `this.services` will not include the service after the kill, whereas `target.services` will always include the service, preventing the network/volume from being removed (since `dependentServices.length === 0` is the prerequisite to kill a network/volume.

Reply:
Another reason `dependentServices` should read from current services instead of target, is that they get union-ed into the set of services to be removed in `compareServices`. It doesn't make sense for target dependent services to be marked for removal since we'd still have to check if they're in current state to avoid a bad kill step.